File: fabric/src/main/java/juuxel/adorn/compat/Compat.java
package juuxel.adorn.compat;

import juuxel.adorn.block.variant.BlockVariantSets;
import juuxel.adorn.config.ConfigManager;
import net.fabricmc.loader.api.FabricLoader;

public final class Compat {
    public static void init() {
        ifModLoaded("biomeswevegone", () -> BlockVariantSets.add(new BwgCompat()));
        ifModLoaded("terrestria", () -> BlockVariantSets.add(new TerrestriaCompat()));
        ifModLoaded("towelette", () -> ToweletteCompat.init());
        ifModLoaded("traverse", () -> BlockVariantSets.add(new TraverseCompat()));
        ifModLoaded("woods_and_mires", () -> BlockVariantSets.add(new WamCompat()));
        ifModLoaded("biomemakeover", () -> BlockVariantSets.add(new BiomeMakeoverCompat()));
        ifModLoaded("cinderscapes", () -> BlockVariantSets.add(new CinderscapesCompat()));
        ifModLoaded("promenade", () -> BlockVariantSets.add(new PromenadeCompat()));
        ifModLoaded("techreborn", () -> BlockVariantSets.add(new TechRebornCompat()));
        ifModLoaded("blockus", () -> BlockVariantSets.add(new BlockusCompat()));
        ifModLoaded("biomesoplenty", () -> BlockVariantSets.add(new BiomesOPlentyCompat()));
    }

    private static void ifModLoaded(String mod, Runnable fn) {
        if (ConfigManager.isCompatEnabled(mod) && FabricLoader.getInstance().isModLoaded(mod)) {
            fn.run();
        }
    }
}

File: fabric/src/main/kotlin/juuxel/adorn/compat/Compat.kt
package juuxel.adorn.compat

import juuxel.adorn.block.variant.BlockVariantSets
import juuxel.adorn.config.ConfigManager
import net.fabricmc.loader.api.FabricLoader

object Compat {
    fun init() {
        val mods = mapOf(
            "byg" to { BlockVariantSets.add(BygCompat) },
            "terrestria" to { BlockVariantSets.add(TerrestriaCompat) },
            "towelette" to ToweletteCompat::init,
            "traverse" to { BlockVariantSets.add(TraverseCompat) },
            "woods_and_mires" to { BlockVariantSets.add(WamCompat) },
            "biomemakeover" to { BlockVariantSets.add(BiomeMakeoverCompat) },
            "cinderscapes" to { BlockVariantSets.add(CinderscapesCompat) },
            "promenade" to { BlockVariantSets.add(PromenadeCompat) },
            "techreborn" to { BlockVariantSets.add(TechRebornCompat) },
            "blockus" to { BlockVariantSets.add(BlockusCompat) },
        )

        for ((mod, fn) in mods) {
            ifModLoaded(mod, fn)
        }
    }

    fun isCompatEnabled(mod: String): Boolean {
        val compatMap = ConfigManager.config().compat

        if (mod !in compatMap) {
            compatMap[mod] = true
            ConfigManager.INSTANCE.save()
            return true
        }

        return compatMap[mod]!!
    }

    private inline fun ifModLoaded(mod: String, fn: () -> Unit) {
        if (isCompatEnabled(mod) && FabricLoader.getInstance().isModLoaded(mod)) {
            fn()
        }
    }
}
